plugins/router/f5/f5.go

```diff
@@ -0,0 +1,1743 @@
+			" HTTP routes will not be configured")
+	}
+
+	if cfg.httpsVserver == "" {
+		glog.Warning("No vserver was specified for HTTPS connections;" +
+			" HTTPS routes will not be configured")
+	}
+
+	if cfg.privkey == "" {
+		glog.Info("No SSH key provided for the F5 BIG-IP host;" +
+			" TLS configuration for applications is disabled")
+	}
+
+	router := &f5LTM{
+		f5LTMCfg: f5LTMCfg{
+			host:         cfg.host,
+			username:     cfg.username,
+			password:     cfg.password,
+			httpVserver:  cfg.httpVserver,
+			httpsVserver: cfg.httpsVserver,
+			privkey:      cfg.privkey,
+			insecure:     cfg.insecure,
+		},
+		poolMembers: map[string]map[string]bool{},
+		routes:      map[string]map[string]bool{},
+	}
+
+	return router, nil
+}
+
+//
+// Helper routines for REST calls.
+//
+
+// One of three things can happen as a result of a REST request to F5
+// iControl:
+//
+// (1) The request succeeds and F5 returns an HTTP 200 response, possibly with
+//     a result payload, which should have the fields defined in the
+//     result argument.  In this case, rest_request decodes the payload into
+//     the result argument and returns nil.
+//
+// (2) The request fails and F5 returns an HTTP 4xx or 5xx response with a
+//     response payload containing a code (which should be the same as the
+//     HTTP response code) and a string message.  In this case, rest_request
+//     decodes the response payload and returns an F5Error with the URL, HTTP
+//     verb, HTTP status code, and error information from the response payload.
+//
+// (3) The REST call fails in some other way, such as a socket error or an
+//     error decoding the result payload.  In this case, rest_request returns
+//     an F5Error with the URL, HTTP verb, HTTP status code (if any), and error
+//     value.
+func (f5 *f5LTM) rest_request(verb string, url string, payload io.Reader,
+	result interface{}) error {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: f5.insecure},
+	}
+
+	errorResult := F5Error{verb: verb, url: url}
+
+	req, err := http.NewRequest(verb, url, payload)
+	if err != nil {
+		errorResult.err = fmt.Errorf("http.NewRequest failed: %v", err)
+		return errorResult
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+	req.SetBasicAuth(f5.username, f5.password)
+
+	client := &http.Client{Transport: tr}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		errorResult.err = fmt.Errorf("client.Do failed: %v", err)
+		return errorResult
+	}
+
+	defer resp.Body.Close()
+
+	errorResult.httpStatusCode = resp.StatusCode
+
+	decoder := json.NewDecoder(resp.Body)
+	if resp.StatusCode != 200 {
+		if resp.StatusCode == 404 {
+			// Just return the error without trying to decode the message
+			// body—F5 sometimes returns an HTML response in the case of a 404
+			// (even though we *ask* for JSON and F5 *usually* obliges).
+			return errorResult
+		}
+		err = decoder.Decode(&errorResult)
```

Comment:
What do you mean? `errorResult` is an `F5Error`, which embeds an `f5Result`, which has the `code` and `message` fields of the JSON response.